wdl-ast/src/experimental/v1/task.rs

```diff
@@ -0,0 +1,1119 @@
+    }
+
+    fn cast(syntax: SyntaxNode) -> Option<Self>
+    where
+        Self: Sized,
+    {
+        match syntax.kind() {
+            SyntaxKind::ParameterMetadataSectionNode => Some(Self(syntax)),
+            _ => None,
+        }
+    }
+
+    fn syntax(&self) -> &SyntaxNode {
+        &self.0
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::experimental::v1::UnboundDecl;
+    use crate::experimental::v1::VisitReason;
+    use crate::experimental::v1::Visitor;
+    use crate::experimental::Document;
+
+    #[test]
+    fn tasks() {
+        let parse = Document::parse(
+            r#"
+version 1.1
+
+task test {
+    input {
+        String name
+    }
+
+    output {
+        String output = stdout()
+    }
+
+    command <<<
+        printf "hello, ~{name}!
+    >>>
+
+    runtime {
+        container: "foo/bar"
+    }
+
+    meta {
+        description: "a test"
+        foo: null
+    }
+
+    parameter_meta {
+        name: {
+            help: "a name to greet"
+        }
+    }
+
+    String x = "private"
+}
+"#,
+        );
+
+        let document = parse.into_result().expect("there should be no errors");
+        let ast = document.ast();
+        let ast = ast.as_v1().expect("should be a V1 AST");
+        let tasks: Vec<_> = ast.tasks().collect();
+        assert_eq!(tasks.len(), 1);
+        assert_eq!(tasks[0].name().as_str(), "test");
+
+        // Task inputs
+        let inputs: Vec<_> = tasks[0].inputs().collect();
+        assert_eq!(inputs.len(), 1);
+
+        // First task input
+        let decls: Vec<_> = inputs[0].declarations().collect();
+        assert_eq!(decls.len(), 1);
+        assert_eq!(
+            decls[0].clone().unwrap_unbound_decl().ty().to_string(),
+            "String"
+        );
+        assert_eq!(
+            decls[0].clone().unwrap_unbound_decl().name().as_str(),
+            "name"
+        );
+
+        // Task outputs
+        let outputs: Vec<_> = tasks[0].outputs().collect();
+        assert_eq!(outputs.len(), 1);
+
+        // First task output
+        let decls: Vec<_> = outputs[0].declarations().collect();
+        assert_eq!(decls.len(), 1);
+        assert_eq!(decls[0].ty().to_string(), "String");
+        assert_eq!(decls[0].name().as_str(), "output");
+        assert_eq!(
+            decls[0]
+                .expr()
+                .unwrap_call()
+                .target()
+                .unwrap_name_ref()
+                .name()
+                .as_str(),
+            "stdout"
+        );
+
+        // Task commands
+        let commands: Vec<_> = tasks[0].commands().collect();
+        assert_eq!(commands.len(), 1);
+
+        // First task command
+        assert!(commands[0].is_heredoc());
+        let parts: Vec<_> = commands[0].parts().collect();
+        assert_eq!(parts.len(), 3);
+        assert_eq!(
+            parts[0].clone().unwrap_text().as_str(),
+            "\n        printf \"hello, "
```

Comment:
At what point will this leading whitespace be stripped? Should that be done by the AST?

Reply:
I think the AST should give the literal text of the command and it's up to a lint rule to check it and an execution engine to format it to whatever script file it's writing out.

Reply:
Ok. Makes sense as the execution engine's responsibility (i.e. the AST does not need to do any whitespace trimming)